### PR TITLE
Added image data luminance source for cases where DOM is not available

### PR DESCRIPTION
**Pull request by Swiftwork**
Created another luminance source for usage with ImageData in web workers and Node.js, but no HTMLCanvas is available. It should also support rotation using shear transforms. However, currently master has linting errors and tests cannot compile so I have trouble verifying the code. Will be a draft until tests can be completed.

### CODE REVIEW

**Comment by odahcam** [comment]
Would you mind transfering this PR to the [browser package](https://github.com/zxing-js/browser)? 

The browser layer is deprecated here and this PR will be merged much faster in the browser package.

Thanks!

**Comment by odahcam** [comment]
Also, this is actually an awesome PR as I was just commenting last month we would need this in the future. This luminance source, as long as I remember, can actually help with decoding from video device streams and ImageCapture API where we currently use canvas for drawing the frame and could just grab the image data buffer instead. ImageData luminace would of course be optional for a while, but I believe this could be the standard luminance for video streams in the future.

**Comment by Swiftwork** [comment]
It might still be a little rough around the edges as it doesn't do the same anti-aliasing that the HTMLCanvas does, but I think the result is very usable. If you want to check out an example I got it here https://codesandbox.io/s/currying-snow-r1431?file=/src/index.js. Regarding moving this to browser project, I am not sure I agree. The use-case I was thinking of is in a web-worker or node.js where no other features from the browser are needed. You might get `ImageData` from other sources than an `HTMLCanvas` or you decode on the server, and requiring installing both libraries seems unnecessary. The actual use-case I am working on is to improve the React Native expo-camera and I wouldn't want to add too much overhead by requiring both dependencies, as I need the entire code. https://github.com/expo/expo/blob/a6d258d6a815169e351f55c357978db78baeb489/packages/expo-camera/src/useWebQRScanner.ts#L39-L44

**Comment by Swiftwork** [comment]
Do you think you will have time to look through the current lint and code tests in the near future so that the Travis CI runs successfully? Would like to add some tests to this code before merging.

**Comment by Swiftwork** [comment]
@odahcam would you be open to using my help to clean up the repository, linting rules, test cases, removal of browser code, etc... If you create a https://gitter.im/ chat for the library we can discuss what needs to be done, and I can do a push this week and the following :)

**Comment by odahcam** [comment]
Yep I would like, unfortunatelly the Gitter kind got unused for two reasons: lack of development conversation and the bad quality of the tool itself. I was thinking recently in creating a Discord server, but we can discuss that somewere else.

**Comment by odahcam** [comment]
And sorry, I do want to merge this but I tought the changes were not ready as the PR status is Draft. Also, I did not proceed with merging or with the conversation because I would have no time to publish and support a new version now.

**Review by odahcam — APPROVED** [review]

**Comment by Swiftwork** [comment]
No worries, when the development branch has a more stable test environment, I will change the PR to point to that branch. I will also add some more tests to this pull request, so no need to merge quite yet.

File: src/test/core/util/AssertUtils.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export default class AssertUtils {
 
 
 export const assertEquals = <T>(actual: any, expected: T, message?: string) => strictEqual(actual, expected, message);
-export const assertArrayEquals = (a: Array<any> | Uint8Array | Int32Array, b: Array<any> | Uint8Array | Int32Array) => deepStrictEqual(a, b);
+export const assertArrayEquals = (a: Iterable<any>, b: Iterable<any>, message?: string) => deepStrictEqual(a, b, message);
```
Comment (odahcam) [inline src/test/core/util/AssertUtils.ts:24]: hahaha killer change! (seriously)

**Review by odahcam — COMMENTED** [review]
I found this changes awesome, I just ask myself if we should start developing these extra functionalities outside of core, which would give us 100% freedom on extending the original lib, and keep the core folder as close as the Java version as possible, that just until we are "behind" the Java version. But this is great, I'd like to see @werthdavid opinion as well.

**Pull request merged**